File: ports/Android/AndroidApp/app/src/main/java/org/junkbot/game/MainActivity.java
package org.junkbot.game;

import android.os.Build;
import android.view.View;
import android.view.WindowManager;
import org.libsdl.app.SDLActivity;

/**
 * SDLActivity (from the vendored SDL3 Java glue, Vendor/java/SDL3-android.jar) drives the whole
 * native lifecycle: it loads the libraries below in order, creates the SDL surface, and invokes
 * libJunkbotAndroid.so's exported {@code SDL_main} on a dedicated thread (which is also what
 * triggers the Swift side's {@code JNI_OnLoad} JVM bootstrap - see
 * ports/Android/Sources/JunkbotAndroid/AndroidMain.swift).
 */
public class MainActivity extends SDLActivity {
    @Override
    protected String[] getLibraries() {
        return new String[] {
            "SDL3",
            "SDL3_image",
            "SDL3_mixer",
            // Last entry doubles as the "main" shared object SDLActivity dlsym's SDL_main from.
            "JunkbotAndroid",
        };
    }

    /**
     * Hides the status/nav bars ("immersive sticky": swiping from an edge reveals them
     * temporarily, then they auto-hide again) so the game truly fills the screen - the
     * Theme.Junkbot manifest theme removes the action bar and lets content draw under the
     * cutout/status bar area, but doesn't hide the system bars themselves.
     *
     * Re-applied on every focus change, not just onCreate: the system bars reappear whenever
     * focus is lost (e.g. a dialog, the recents switcher) and won't rehide on their own when
     * focus returns.
     */
    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemBars();
        }
    }

    private void hideSystemBars() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            getWindow().setDecorFitsSystemWindows(false);
            android.view.WindowInsetsController controller = getWindow().getInsetsController();
            if (controller != null) {
                controller.hide(android.view.WindowInsets.Type.statusBars()
                    | android.view.WindowInsets.Type.navigationBars());
                controller.setSystemBarsBehavior(
                    android.view.WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
            }
        } else {
            getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_FULLSCREEN
                | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
        }
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }
}

File: ports/Android/AndroidApp/app/src/main/java/org/junkbot/game/MainActivity.kt
package org.junkbot.game

import android.app.Activity
import android.os.Bundle

// Not verified - see ../../../../../README.md's "target type" gap. Real SDL3 Android apps
// normally extend SDL's own SDLActivity (from libsdl-org/SDL's android-project template, which
// bootstraps SDL_main on a dedicated thread and wires up the GL/Vulkan surface) rather than a
// plain Activity - that template was not vendored here since doing so meaningfully requires
// testing against a real build. Whichever entry point JunkbotSDL3's JNI code ends up exposing
// (see README.md) will need to be called from here, or this should extend SDLActivity instead
// once that's vendored in.
class MainActivity : Activity() {
    companion object {
        init {
            System.loadLibrary("JunkbotSDL3")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // TODO: call into the native library's actual entry point once ports/Android's
        // Package.swift target-type gap (executable vs. dynamic library + JNI_OnLoad) is
        // resolved - loadLibrary above only maps the .so into the process, it doesn't run
        // anything yet.
    }
}
